[material] Remove dependency to @mui/base

mnajdova opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [ ] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

aarongarciah commented:
@mnajdova today I merged https://github.com/mui/material-ui/pull/37301, which contains changes in mui-base. Can you update this PR to include them?

mnajdova commented:
> @mnajdova today I merged https://github.com/mui/material-ui/pull/37301, which contains changes in mui-base. Can you update this PR to include them?

Thanks for the heads up! I will update it tomorrow and have this ready for review :)

aarongarciah commented on docs/pages/material-ui/api/popper.json:
```diff
@@ -67,7 +67,14 @@
-  "classes": [],
+  "classes": [
+    {
+      "key": "root",
```
Why didn't it have a class name before?

mnajdova replied:
It was coming from Base UI, I think the docs didn't picked it up.

aarongarciah reviewed (approved):
Looks good

mnajdova commented on packages/api-docs-builder-core/baseUi/projectSettings.ts:
```diff
@@ -27,6 +27,9 @@ export const projectSettings: ProjectSettings = {
   translationLanguages: LANGUAGES,
   skipComponent: () => false,
+  skipHook: (filename) => {
+    return filename.match(/(useSlotProps)/) !== null;
```
This hook was anyway not documented, so I've added this option to skip, otherwise the docgen generation failed. It's an optional prop it won't break other ts projects.

mnajdova commented:
@aarongarciah this is ready for final review. Changes after the approval:
- moved the utils used in MUI X to @mui/utils to avoid having to add @mui/material as a dependency to some packages that are Material UI agnostic (charts for example)ho
- fixed the docs generation script, as Base UI is not just re-exporting some hooks from @mui/utils (useSlotProps)

aarongarciah reviewed (approved):
🏎️ 🚀

mnajdova commented:
Thanks for the review @aarongarciah. Next steps:
- cherry-pick on master
- create PR on MUI X with the updates
- after we ensure everything works, mark the @mui/base package as private and stop releasing it

I will move with the next steps.

oliviertassinari commented on docs/data/material/pagesApi.js:
```diff
@@ -27,6 +27,7 @@ module.exports = [
   { pathname: '/material-ui/api/checkbox' },
   { pathname: '/material-ui/api/chip' },
   { pathname: '/material-ui/api/circular-progress' },
+  { pathname: '/material-ui/api/click-away-listener' },
```
I'm confused about this. Would it mean we will have:

- `@mui/material/ClickAwayListener`
- `@mui/base/ClickAwayListener`
- `@base_ui/react/ClickAwayListener`. I mean, I imagine we need it https://stackoverflow.com/questions/32553158/detect-click-outside-react-component 

<img width="512" alt="SCR-20240713-bcxd" src="https://github.com/user-attachments/assets/bcd23fff-e423-40f7-bc03-e06266fcbe73">

oliviertassinari commented on docs/data/material/pagesApi.js:
```diff
@@ -125,6 +128,7 @@ module.exports = [
   { pathname: '/material-ui/api/tab-panel' },
   { pathname: '/material-ui/api/tabs' },
   { pathname: '/material-ui/api/tab-scroll-button' },
+  { pathname: '/material-ui/api/textarea-autosize' },
```
Same idea as above with https://stackoverflow.com/questions/454202/creating-a-textarea-with-auto-resize

<img width="541" alt="SCR-20240713-bdjc" src="https://github.com/user-attachments/assets/d386a9d3-d6ed-4df3-9720-b163bbbfb014">

oliviertassinari reviewed (commented):
What's the motivation for this change? It seems that the PR description is missing context.

oliviertassinari commented:
We miss the history on those components after this PR. From https://github.com/mui/mui-x/pull/13823 I think that we should revert this PR.

mnajdova commented:
> From the exposure to this change and dive discussed in https://github.com/mui/mui-x/pull/13823 can we revert this PR and go in the opposite direction?

What exactly is the opposite direction?

> I also got confused in https://github.com/mui/material-ui/issues/42604#issuecomment-2237645100, we shouldn't have 3 versions at any point in time. As a side note, we need the git history of those components, it's missing after this PR.

Right, I would assume by moving the component to Mateiral UI, we won't need to do any updates on @mui/base, the initial plan is to no longer release this package, mark it as private, and eventually remove the code once we have @base_ui.

> It seems that having those modules in @mui/base, hosted in [http://github.com/mui/material-ui](https://github.com/mui/material-ui) would already to be a clear step forward, with no real downsides.

The motivation for the change is that we will no longer need to make updates on @mui/base, not about where the components are being documented.

> Now, I'm truly advocating to have those hosted in [http://github.com/mui/base-ui](https://github.com/mui/base-ui) and exposed from @base_ui/react as unstable/legacy.

I am not sure which if any of these components will live in Base UI, and what would be the form of these. What is the value of adding legacy stuff in Base UI, if these are only used in Material UI and Base UI do not plan to support these or plans to change the API around them. I will let @colmtuite and @michaldudak answer on this, but in my opinion this was not an option, hence I went with moving them to Material UI.

mnajdova commented:
> As a side note, we need the git history of those components, it's missing after this PR.

I can fix this in https://github.com/mui/material-ui/pull/43028, I've applied the changes needed for the NoSsr, but I will wait with updating all components until we have a decision on how to move forward.

colmtuite commented:
> I would assume by moving the component to Mateiral UI, we won't need to do any updates on @mui/base, the initial plan is to no longer release this package, mark it as private, and eventually remove the code once we have @base_ui.

This makes sense to me, and is what the majority of the team have been planning for at least the past few months. I figure we would just take each util/package, figure out the most appropriate place for it to belong, and move it there. Then that team are responsible for it. Sometimes, the most appropriate place will be Base UI, sometimes, Material UI, sometimes X, and potentially sometimes a shared utils lib.

But corrupting the new Base UI lib by prematurely abstracting legacy utils/components to it—when we don't know if they belong there long-term—is the worst path we could take. It's much worse than duplicating things in multiple locations.
